File: app/src/main/java/de/gaffga/android/fragments/SessionTouchHelperCallback.java
package de.gaffga.android.fragments;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

public class SessionTouchHelperCallback extends ItemTouchHelper.Callback {

    private final SessionTouchListener listener;

    public interface SessionTouchListener {
        boolean onMove(int fromPosition, int toPosition);
    }

    public SessionTouchHelperCallback(SessionTouchListener listener) {
        this.listener = listener;
    }

    @Override
    public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
        int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        return makeMovementFlags(dragFlags, 0);
    }

    @Override
    public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
        if (listener != null) {
            return listener.onMove(viewHolder.getBindingAdapterPosition(), target.getBindingAdapterPosition());
        }
        return false;
    }

    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return true;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return false;
    }
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/fragments/SessionTouchHelperCallback.kt
package at.priv.graf.zazentimer.fragments

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

class SessionTouchHelperCallback(
    private val listener: SessionTouchListener?,
) : ItemTouchHelper.Callback() {
    interface SessionTouchListener {
        fun onSwipe(position: Int)

        fun onMove(
            fromPosition: Int,
            toPosition: Int,
        ): Boolean

        fun onDragEnd()
    }

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
    ): Int {
        val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
        val swipeFlags = ItemTouchHelper.START
        return makeMovementFlags(dragFlags, swipeFlags)
    }

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder,
    ): Boolean = listener?.onMove(viewHolder.bindingAdapterPosition, target.bindingAdapterPosition) ?: false

    override fun onSwiped(
        viewHolder: RecyclerView.ViewHolder,
        direction: Int,
    ) {
        listener?.onSwipe(viewHolder.bindingAdapterPosition)
    }

    override fun clearView(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
    ) {
        listener?.onDragEnd()
    }

    override fun isLongPressDragEnabled(): Boolean = false

    override fun isItemViewSwipeEnabled(): Boolean = true
}
